lib/exhal/form_field.ex

```diff
@@ -0,0 +1,93 @@
+defmodule ExHal.FormField do
+  @moduledoc """
+  Functions for working with HAL form fields.
+  """
+
+  @typedoc """
+  A HAL form field.
+  """
+  @type t :: %__MODULE__{
+          name: String.t(),
+          type: field_types,
+          path: String.t(),
+          value: field_value
+        }
+  defstruct [
+    :name,
+    :type,
+    :path,
+    :value
+  ]
+
+  @typedoc """
+  Field types.
+  """
+  @type field_types ::
+          :boolean
+          | :string
+          | :number
+          | :date
+          | :time
+          | :datetime
+          | :sensitive
+          | :hidden
+          | :text
+          | :email
+          | :tel
+          | :file
+
+  @typedoc """
+  Possible value types.
+  """
+  @type field_value :: nil | true | false | list | float | integer | String.t()
+
+  @spec from_field_entry(%{}) :: __MODULE__.t()
+  @doc """
+  Create a new form field from parsed form json.
+
+  Raises `ArgumentError` if json is invalid.
+  """
+  def from_field_entry(a_map) do
+    %__MODULE__{
+      name: extract_name(a_map),
+      type: extract_type(a_map),
+      path: extract_path(a_map),
+      value: extract_value(a_map)
+    }
+  end
+
+  @spec set_value(__MODULE__.t(), field_value) :: __MODULE__.t()
+  @doc """
+  Returns a form field with the specified value.
+
+  Raises `ArgumentError` if the new value is the wrong type.
```

Comment:
I don't see how this can raise. Am I missing something?

Reply:
It will someday. :)